cpp/oneapi/dal/table/backend/interop/host_homogen_table_adapter.cpp

```diff
@@ -81,26 +82,62 @@ auto host_homogen_table_adapter<Data>::create(const homogen_table& table) -> ptr
 
 template <typename Data>
 host_homogen_table_adapter<Data>::host_homogen_table_adapter(const homogen_table& table,
-                                                             status_t& status)
-        // The following const_cast is safe only when this class is used for read-only
-        // operations. Use on write leads to undefined behaviour.
-        : base(daal_dm::DictionaryIface::equal,
-               ptr_data_t{ const_cast<Data*>(table.get_data<Data>()), daal_object_owner(table) },
-               table.get_column_count(),
-               table.get_row_count(),
-               status),
-          is_rowmajor_(table.get_data_layout() == data_layout::row_major),
+                                                             status_t& stat)
+        : NumericTable(dal::detail::integral_cast<std::size_t>(table.get_column_count()),
+                       dal::detail::integral_cast<std::size_t>(table.get_row_count())),
           original_table_(table) {
-    if (!status.ok()) {
+    // The following const_casts are safe only when this class is used for read-only
+    // operations. Use on write leads to undefined behaviour.
+    if (!stat.ok()) {
         return;
     }
     else if (!table.has_data()) {
-        status.add(daal::services::ErrorIncorrectParameter);
+        stat.add(daal::services::ErrorIncorrectParameter);
         return;
     }
 
+    const std::size_t column_count =
+        dal::detail::integral_cast<std::size_t>(table.get_column_count());
+    const std::size_t row_count = dal::detail::integral_cast<std::size_t>(table.get_row_count());
+
+    if (table.get_data_layout() == data_layout::row_major) {
+        const auto original_data = const_cast<Data*>(table.get_data<Data>());
```

Comment:
Pull it up before the condition check and use `origin_data` instead `const_cast<Data*>(&(table.get_data<Data>()[i * row_count]));` in the SOA branch

Reply:
applied